syntax/grammar.mjs

```diff
@@ -376,14 +376,39 @@ let Function =
 /* Content Characters
  *
  * Translation content can be written using most Unicode characters, with the
- * exception of C0 control characters (but allowing tab), surrogate blocks and
- * non-characters (U+FFFE, U+FFFF).
+ * exception of C0 and C1 control characters (but including TAB), DEL,
+ * surrogate blocks and non-characters. New lines are excluded because they are
+ * forbidden in quoted_text and handled with additional logic in block_text.
  */
 
 let any_char =
     either(
-        charset("\\u{9}\\u{20}-\\u{D7FF}\\u{E000}-\\u{FFFD}"),
-        charset("\\u{10000}-\\u{10FFFF}"));
+        // TAB
+        charset("\\u{9}"),
+        // Basic Latin.
+        charset("\\u{20}-\\u{7E}"),
+        // Skip DEL and C1 controls.
+        // Resume at Latin-1.
+        charset("\\u{A0}-\\u{D7FF}"),
+        // Skip surrogate codepoints.
+        // Resume at the U+E000 PUA.
+        charset("\\u{E000}-\\u{FDCF}"),
+        // Skip the non-character block U+FDD0..FDEF */
+        // Resume at U+FDF0.
+        charset("\\u{FDF0}-\\u{FFFD}"),
```

Comment:
I tried adding a test with `abc<U+D800>def` (the first high surrogate) and somehow, this `charset` parser successfully parses it. I'm not sure what's going on yet.

Reply:
OK, I think I figured out why this was so. In UTF-8, surrogate pair as well as lonely surrogates are ill-formed. Conformant decoders must handle it in some way:

> When faced with this illegal byte sequence while transforming or interpreting, a UTF-8 conformant process must treat the first byte [of the example] as an illegal termination error: for example, either signaling an error, filtering the byte out, or representing the byte with a marker such as FFFD (REPLACEMENT CHARACTER). <sub>[source](http://unicode.org/faq/utf_bom.html#gen8)</sub>

It looks like node defaults to using the `U+FFFD` (�) character in such cases.  And that's exaclt the character the parser sees here.

To write a test for rejecting surrogates in the parser, we'd need to have a different test runner which would build a test string in JavaScript, rather than reading it from a file using the UTF-8 encoding.

Reply:
I think we should rely on UTF-8 protecting us here.